demos/defender/defender_demo_json/mqtt_operations.c

```diff
@@ -1009,17 +1118,30 @@ bool PublishToTopic( const char * pTopicFilter,
 }
 /*-----------------------------------------------------------*/
 
-bool ProcessLoop( uint32_t timeoutMs )
+bool ProcessLoopWithTimeout( uint32_t ulTimeoutMs )
 {
+    uint32_t ulMqttProcessLoopTimeoutTime;
+    uint32_t ulCurrentTime;
+
+    MQTTStatus_t eMqttStatus = MQTTSuccess;
     bool returnStatus = false;
-    MQTTStatus_t mqttStatus = MQTTSuccess;
 
-    mqttStatus = MQTT_ProcessLoop( &mqttContext, timeoutMs );
+    ulCurrentTime = mqttContext.getTime();
+    ulMqttProcessLoopTimeoutTime = ulCurrentTime + ulTimeoutMs;
 
-    if( mqttStatus != MQTTSuccess )
+    /* Call MQTT_ProcessLoop multiple times until the timeout expires or
+     * #MQTT_ProcessLoop fails. */
+    while( ( ulCurrentTime < ulMqttProcessLoopTimeoutTime ) &&
+           ( eMqttStatus == MQTTSuccess || eMqttStatus == MQTTNeedMoreBytes ) )
+    {
+        eMqttStatus = MQTT_ProcessLoop( &mqttContext );
+        ulCurrentTime = mqttContext.getTime();
```

Comment:
Here as well, can we add a small delay here?

Reply:
I don't have the run-time context here, so my comments may be irrelevant:  If this is busy waiting in a multithreaded system at a priority above that of any part of the network stack then there is risk of deadlock.  However, it is also not good for responsiveness and throughput to put the thread/task to sleep for an arbitrary period of time.  Hence, if you do put the thread to sleep to wait, unblock on an event, rather than a timeout (actually, unblock on both, in case the event never arrives).   On the other hand, if this is not multithreaded, and there is nothing that can be done until the packet arrives, power saving would be the only gain my adding a delay.

Reply:
The demo is running on one thread - there is no need to add a delay here... as it will just stall the single task. deadlock is avoided by the timeout. This function is essentially supposed to just work as the old MQTT_ProcessLoop did for places where the demo architecture relied on it to wait for incoming publishes. prvWaitForPacket is used in places where the demo waits on an ack.